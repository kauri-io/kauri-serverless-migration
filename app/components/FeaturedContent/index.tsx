import styled from 'styled-components'
import ResourceRow from '../SearchResults/ResourceRowWithImage' // IProps as ResourceRowProps,
import { Title2 } from '../Typography'
import { RenderCardContent } from '../CuratedLists'
import theme from '../../lib/theme-config'

const DEFAULT_CARD_WIDTH = theme.DEFAULT_CARD_WIDTH

const Container = styled.section`
    display: flex;
    flex-direction: column;
    padding: 0px;
    > :first-child {
        margin-bottom: ${props => props.theme.space[2]}px;
    }
    @media (max-width: ${props => props.theme.breakpoints[2]}) {
        > h3 {
            padding: 0px ${props => props.theme.space[2]}px;
        }
    }
`

const ResourceContainer = styled(Container)`
    padding: 0px;
    > :not(:last-child) {
        margin-bottom: ${props => props.theme.space[2]}px;
    }
    @media (max-width: ${props => props.theme.breakpoints[2]}) {
        align-items: center;
    }
`

const DesktopResourceContainer = styled(ResourceContainer)`
    @media (max-width: ${props => props.theme.breakpoints[2]}) {
        display: none;
    }
`

const MobileResourceContainer = styled(ResourceContainer)`
    @media (max-width: ${props => props.theme.breakpoints[2]}) {
        display: grid;
        width: 100vw;
        > :not(:first-child) {
            justify-content: center;
        }
        grid-template-columns: repeat(
            auto-fill,
            minmax(${DEFAULT_CARD_WIDTH}px, ${DEFAULT_CARD_WIDTH}px)
        );
        grid-gap: ${props => props.theme.space[2]}px
            ${props => props.theme.space[3]}px;
    }
    @media (max-width: ${props => props.theme.breakpoints[0]}) {
        justify-content: center;
    }
    @media (min-width: ${props => props.theme.breakpoints[2]}) {
        display: none;
    }
`

interface IProps {
    content: any[]
    Link: React.ReactNode
}

const RenderMobileFeaturedContent: React.FunctionComponent<IProps> = props => {
    return (
        <MobileResourceContainer>
            {props.content.map(({ resource }) => {
                const articleCount =
                    resource.sections &&
                    resource.sections.reduce((current: number, next: any) => {
                        if (next && Array.isArray(next.resources)) {
                            const articlesInSection = next.resources.filter(
                                (sectionResource: any) => {
                                    return (
                                        sectionResource &&
                                        sectionResource.__typename
                                            .toLowerCase()
                                            .includes('article')
                                    )
                                }
                            )
                            return articlesInSection.length + current
                        }
                        return current
                    }, 0)

                const collectionCount =
                    resource.sections &&
                    resource.sections.reduce((current: number, next: any) => {
                        if (next && Array.isArray(next.resources)) {
                            const collectionsInSection = next.resources.filter(
                                (sectionResource: any) =>
                                    sectionResource &&
                                    sectionResource.__typename
                                        .toLowerCase()
                                        .includes('collection')
                            )
                            return collectionsInSection.length + current
                        }
                        return current
                    }, 0)

                const RenderedCard = RenderCardContent({
                    Link: props.Link,
                    fromAdmin: false,
                })

                return (
                    <RenderedCard
                        {...resource}
                        articleCount={articleCount}
                        collectionCount={collectionCount}
                    />
                )
            })}
        </MobileResourceContainer>
    )
}

const RenderDesktopFeaturedContent: React.FunctionComponent<
    IProps & { Link: React.FunctionComponent<any> }
> = ({ content, Link }) => (
    <DesktopResourceContainer>
        {content.map(({ resource }: { resource: any }) => {
            switch (resource.__typename) {
                case 'ArticleDTO': {
                    return (
                        <ResourceRow
                            resourceType={resource.__typename
                                .split('DTO')[0]
                                .toLowerCase()}
                            key={String(resource.id)}
                            id={String(resource.id)}
                            version={Number(resource.version)}
                            title={String(resource.title)}
                            description={String(resource.description)}
                            tags={
                                (resource.tags &&
                                    resource.tags.map((tag: string) =>
                                        tag ? tag : ''
                                    )) ||
                                []
                            }
                            username={
                                (resource.owner as any).username ||
                                (resource.owner as any).name
                            }
                            userId={(resource.owner as any).id}
                            userAvatar={(resource.owner as any).avatar}
                            date={resource.datePublished}
                            ownerType={'USER'}
                            linkComponent={(children, route) => (
                                <Link useAnchorTag={true} href={route}>
                                    {children}
                                </Link>
                            )}
                            imageURL={
                                resource &&
                                resource.attributes &&
                                resource.attributes.background
                            }
                        />
                    )
                }

                case 'CollectionDTO': {
                    return (
                        <ResourceRow
                            resourceType={resource.__typename
                                .split('DTO')[0]
                                .toLowerCase()}
                            key={String(resource.id)}
                            id={String(resource.id)}
                            title={String(resource.name)}
                            description={String(resource.description)}
                            tags={
                                (resource.tags &&
                                    resource.tags.map((tag: string) =>
                                        tag ? tag : ''
                                    )) ||
                                []
                            }
                            username={
                                (resource.owner as any).username ||
                                (resource.owner as any).name
                            }
                            userId={(resource.owner as any).id}
                            userAvatar={(resource.owner as any).avatar}
                            date={resource.dateUpdated}
                            ownerType={'USER'}
                            linkComponent={(children, route) => (
                                <Link useAnchorTag={true} href={route}>
                                    {children}
                                </Link>
                            )}
                            imageURL={resource.background}
                        />
                    )
                }

                default: {
                    return null
                }
            }
        })}
    </DesktopResourceContainer>
)

const FeaturedContent: React.FunctionComponent<IProps> = props => (
    <Container>
        <Title2>Featured Content</Title2>
        <RenderDesktopFeaturedContent
            {...props}
            Link={props.Link as React.FunctionComponent<any>}
        />
        <RenderMobileFeaturedContent {...props} />
    </Container>
)

export default FeaturedContent
