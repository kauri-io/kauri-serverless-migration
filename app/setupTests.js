import Enzyme, { shallow, mount } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import theme from './lib/theme-config'
import { ThemeProvider } from 'styled-components'

Enzyme.configure({ adapter: new Adapter() })

const observeMock = {
    observe: () => null,
    disconnect: () => null, // maybe not needed
}

window.IntersectionObserver = () => observeMock

export const mountWithTheme = children =>
    mount(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

export const renderWithTheme = children =>
    render(<ThemeProvider theme={theme}>{children}</ThemeProvider>)

export const shallowWithTheme = children =>
    shallow(<ThemeProvider theme={theme}>{children}</ThemeProvider>)
