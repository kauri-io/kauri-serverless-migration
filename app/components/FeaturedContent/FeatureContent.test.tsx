import FeaturedContent from './index'
import { shallowWithTheme } from '../../setupTests'

let props = {
    content: [
        {
            resource: [
                {
                    id: 'test',
                    __typename: 'ArticleDTO',
                },
                {
                    id: 'test2',
                    __typename: 'CollectionDTO',
                },
            ],
        },
    ],
    Link: jest.fn(),
}

describe('components/FeaturedContent', () => {
    it('snapshot should match', () => {
        const wrapper = shallowWithTheme(<FeaturedContent {...props} />)
        expect(wrapper).toMatchSnapshot()
    })
})
