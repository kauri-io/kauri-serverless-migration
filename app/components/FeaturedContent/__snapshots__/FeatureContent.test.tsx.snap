// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/FeaturedContent snapshot should match 1`] = `
<ContextConsumer>
  <Component />
</ContextConsumer>
`;
